spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -201,9 +202,20 @@ case class DeltaSharingSource(
   private def maybeGetFileChanges(
       fromVersion: Long,
       fromIndex: Long,
-      isStartingVersion: Boolean): Unit = {
+      isStartingVersion: Boolean,
+      endOffsetOption: Option[DeltaSharingSourceOffset]): Unit = {
     if (!sortedFetchedFiles.isEmpty) {
-      return
+      val headFile = sortedFetchedFiles.head
+      if ((headFile.version > fromVersion || (headFile.version == fromVersion && headFile.index >
+        fromIndex && fromIndex != -1)) && headFile.isSnapshot && endOffsetOption.isDefined) {
+        // clean up local sortedFileIndex, re-fetch files.
+        // This happens when the fall back logic of AvailableNow calls the latestOffset first
+        // on a query where "startingVersion" is not specified, it gets the latest table snapshot,
+        // which contains files not matching getBatch.
+        sortedFetchedFiles = Seq.empty
+      } else {
+        return
+      }
```

Comment:
Question: Would we support Trigger.AvailableNow after this change?

Reply:
I think it's still better to keep the check to make sure we safely reuse `sortedFetchedFiles`.